hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE NoStarIsType            #-}
+{-# LANGUAGE OverloadedLists         #-}
+{-# LANGUAGE PolyKinds               #-}
+{-# LANGUAGE RankNTypes              #-}
+{-# LANGUAGE RecordWildCards         #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+{-# LANGUAGE TypeApplications        #-}
+{-# LANGUAGE TypeFamilies            #-}
+{-# LANGUAGE TypeOperators           #-}
+{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.KnownNat.Solver #-}
+{-# OPTIONS_GHC -fplugin GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=0 #-}
+
+module Torch.Typed.NN.Recurrent.LSTM where
+
+import           Control.Monad                  ( foldM
+                                                , when
+                                                , void
+                                                )
+import           Control.Exception.Safe         ( try
+                                                , SomeException(..)
+                                                )
+import           Data.List                      ( foldl'
+                                                , intersperse
+                                                , scanl'
+                                                )
+import           Data.Maybe                     ( catMaybes )
+import           Data.Reflection
+import           Foreign.ForeignPtr
+import           GHC.Generics
+import           GHC.TypeLits
+import           GHC.TypeLits.Extra
+import           System.Environment
+import           System.IO.Unsafe
+import           Prelude                 hiding ( tanh )
+import qualified ATen.Cast                     as ATen
+import qualified ATen.Class                    as ATen
+import qualified ATen.Type                     as ATen
+import qualified ATen.Managed.Type.Tensor      as ATen
+import qualified Torch.Autograd                as A
+import qualified Torch.DType                   as D
+import qualified Torch.Functions               as D
+import qualified Torch.NN                      as A
+import qualified Torch.Tensor                  as D
+import qualified Torch.TensorFactories         as D
+import           Torch.Typed
+import           Torch.Typed.Factories
+import           Torch.Typed.Native      hiding ( linear )
+import           Torch.Typed.NN
+
+
+
+data LSTMCellSpec (dtype :: D.DType) (inputDim :: Nat) (hiddenDim :: Nat) =
+    LSTMCellRandomInitState -- ^ Random initial memory cells and hiddenstate values
+    | LSTMCellZeroInitState -- ^ Zero-valued initial memory cells and hiddenstate values
+    deriving (Show, Eq, Ord, Generic, Enum, Bounded)
+
+data LSTMCell (dtype :: D.DType) (inputDim :: Nat) (hiddenDim :: Nat) =  LSTMCell {
+    lSTMCell_init_h :: Parameter dtype '[hiddenDim] -- ^ initial hidden state
+    , lSTMCell_init_c :: Parameter dtype '[hiddenDim] -- ^ initial cell state
```

Comment:
Why do we keep the initial states in the data structure? We should probably reallocate them at each step.

Reply:
I concur. The initial states are not usually not trainable, but are supplied at the beginning of each recurrence.

Reply:
Since the lstmCell does not manage recurrence, it expects the initial state to be provided and whatever the provided values are should be kept as part of the state of the cell; pytorch [defaults to zeros](https://pytorch.org/docs/stable/nn.html#lstmcell).  

The `foldl'`-based `forward` function threads the accumulated state through the container and passes the current state through each cell transition.

I had thought somewhere in the docs I saw that these were trainable, but it looks like although sensible [this is far from canonical](https://r2rt.com/non-zero-initial-states-for-recurrent-neural-networks.html).  I'll expose this case as a separate constructor.

Reply:
I'm very confused now. I think we have to be careful not to mix up rnns and rnn cells.
In https://pytorch.org/docs/stable/_modules/torch/nn/modules/rnn.html, you can see two groups of implementations, one for rnns (base rnn, gru, lstm) and one for rnn cells (base rnn cell, gru cell, lstm cell).
In the latter case, for the rnn cells, not only the tensor input to the cell but also a hidden state is provided.
In the former case, for the rnns the module makes an assumption of what the first hidden state should be.
I think you made this assumption for the rnn cells where you should have made that assumption only for the rnns.

Thus, either don't make any assumptions for your cell implementation, or add new classes for lstm rnns.

Reply:
> I think you made this assumption for the rnn cells where you should have made that assumption only for the rnns.

That's a fair assessment;  It seems like it makes sense forgo any default (hidden,cell) state for the `_cell` functions.  However, this wouldn't conform to the expectations of a pytorch user.

Reply:
I think we agree that the hidden states of cells are meant to be fed into the same cell again, over and over again, as one is folding the input sequence. It therefore does not make sense to feed in a default value at every step.
This begs the question what the first input hidden state should be. Pytorch does not have an opinion on that at the level of cells. The forward function of a cell is just one step of the recurrence and hence always takes both, input step features and hidden state.

Reply:
My suggestion would be that you follow the approach of Pytorch add another class, LSTMRNN. This one would indeed be opinionated on the first input hidden state, because its forward function is calling the lstm cell recurrently

Reply:
> add another class, LSTMRNN [...] would indeed be opinionated [...]

Yes, that was my desired end-goal.  After the feedback, I think that this PR should only scope over the `LSTMCell` implementation and the other components should be separate.

> Pytorch does not have an opinion on that at the level of cells.

 [It does](https://pytorch.org/docs/stable/nn.html#lstmcell): if the current hidden and cell states are not provided to `forward`, they are set to zero values. But Aten makes no such assumptions and fails at runtime if two states are not passed.  This is the ambiguity that led to the current implementation.  

To be clear:  Your preference is that `lstm_cell` should always be provided `h_0` and `c_0` (and defaults should only be the prerogative of the code calling `lstm_cell`), contrary to the above PyTorch api?  This seems to me to be the cleaner api boundary, but, again, totally your call.

Reply:
> It does: ...

Thanks for pointing that out, I wasn't aware of that.

> Your preference is that lstm_cell should always be provided h_0 and c_0 (and defaults should only be the prerogative of the code calling lstm_cell), contrary to the above PyTorch api?

Yes, it would be my personal preference.